### PR TITLE
Update spring boot dependency version to 2.4.0 and spring cloud to 2020.0.0

### PR DESCRIPTION
**Pull request by yiliuTo**
Update version of spring-boot-dependencies to 2.4.0 and spring-cloud-dependencies to 2020.0.0

### CODE REVIEW

**Comment by chenrujun** [comment]
Hi, @yiliuTo .

Please fix the pipeline failure.
After that, please run `/azp java - spring - tests`

**Comment by yiliuTo** [comment]
Hi @jianghaolu , we are updating some dependency libraries to follow the [versions of Spring Boot dependencies](https://repo.maven.apache.org/maven2/org/springframework/boot/spring-boot-dependencies/2.4.0/), and it touches your code in azure-core-amqp and azure-core-http-netty. We want to update **io.projectreactor.netty: reactor-netty** from 0.9.13.RELEASE to 1.0.1, but it has breaking changes and causes the com.azure.core.http.netty.NettyAsyncHttpClientBuilder cannot work, so would you like to take a look and handle it?
Thanks

**Comment by mitchdenny** [comment]
@JonathanGiles

**Comment by bilak** [comment]
Are you aware of that spring boot 2.4 is [not compatible](https://spring.io/projects/spring-cloud) with spring cloud Hoxton?

**Comment by Saljack** [comment]
Hi I just want to inform you that Spring Cloud Hoxton.SR9 is compatible only with Spring Boot 2.3.x and 2.2.x. so you should upgrade Spring Cloud to 2020.0 (Ilford).  See [Spring Cloud project page](https://spring.io/projects/spring-cloud) and compatibility table there.

File: eng/versioning/external_dependencies.txt [hunk 1]
```diff
@@ -92,56 +92,59 @@ org.slf4j:slf4j-nop;1.7.30
...
+org.springframework:spring-web;5.3.1
+org.springframework:spring-webmvc;5.3.1
 
-# spring-boot-starter-parent is not managed by spring-boot-dependencies or spring-cloud-dependencies.
-org.springframework.boot:spring-boot-starter-parent;2.3.7.RELEASE
+# spring-boot-starter-parent's version should be updated manually. Because:
+# 1. It's not managed by spring-boot-dependencies or spring-cloud-dependencies.
+# 2. When spring-boot have minor version up, we should update it's version. But should not
+#    update spring-boot-dependencies' version.
+org.springframework.boot:spring-boot-starter-parent;2.4.0
```
Comment (chenrujun) [inline eng/versioning/external_dependencies.txt:139]: @yiliuTo , FYI.

File: eng/versioning/external_dependencies.txt [hunk 2]
```diff
@@ -33,31 +33,31 @@ com.microsoft.azure:azure-servicebus;3.4.0
 com.microsoft.azure:qpid-proton-j-extensions;1.2.3
 com.microsoft.rest:client-runtime;1.7.4
 com.microsoft.rest.v2:client-runtime;2.1.1
-com.microsoft.sqlserver:mssql-jdbc;7.4.1.jre8
+com.microsoft.sqlserver:mssql-jdbc;8.4.1.jre8
```
Comment (chenrujun) [inline eng/versioning/external_dependencies.txt:36]: @alzimmermsft, FYI: It's another major version up.

File: eng/versioning/external_dependencies.txt [hunk 3]
```diff
@@ -213,8 +216,8 @@ uk.org.lidalia:slf4j-test;1.2.0
 com.azure:sdk-build-tools;1.0.0
 com.beust:jcommander;1.78
 com.google.code.findbugs:jsr305;3.0.2
-com.nimbusds:nimbus-jose-jwt;8.19
-com.nimbusds:oauth2-oidc-sdk;7.1.1
+com.nimbusds:nimbus-jose-jwt;9.1.2
+com.nimbusds:oauth2-oidc-sdk;8.23.1
```
Comment (chenrujun) [inline eng/versioning/external_dependencies.txt:220]: @alzimmermsft , Same here.

File: sdk/spring/scripts/update_spring_managed_external_dependencies.py [hunk 4]
```diff
@@ -60,7 +61,6 @@ def update_dependency_dict(dependency_dict, root_pom_id):
     )
     q = queue.Queue()
     q.put(root_pom)
-    visited_pom_dict = {root_pom_group_artifact:root_pom_version}
```
Comment (chenrujun) [inline sdk/spring/scripts/update_spring_managed_external_dependencies.py:63]: We already have `dependency_dict`, do I deleted `visited_pom_dict`

File: sdk/spring/scripts/update_spring_managed_external_dependencies.py [hunk 5]
```diff
@@ -103,21 +103,17 @@ def update_dependency_dict(dependency_dict, root_pom_id):
                 dependency_dict[key] = version
                 log.debug('Dependency version added. key = {}, value = {}'.format(key, version))
             elif version != dependency_dict[key]:
-                log.debug('Dependency version skipped. key = {}, version = {}, dependency_dict[key] = {}.'.format(key, version, dependency_dict[key]))
+                log.info('Dependency version skipped. key = {}, version = {}, dependency_dict[key] = {}.'.format(key, version, dependency_dict[key]))
             artifact_type = dependency_element.find('./maven:type', MAVEN_NAME_SPACE)
             artifact_scope = dependency_element.find('./maven:scope', MAVEN_NAME_SPACE)
-            if all([artifact_type, artifact_scope]) and artifact_type.text.strip() == 'pom' and artifact_scope.text.strip() == 'import':
```
Comment (chenrujun) [inline sdk/spring/scripts/update_spring_managed_external_dependencies.py:109]: `all(...)` does not work well.

**Review by chenrujun — COMMENTED** [review]
@alzimmermsft , I think you can merge this branch to your branch.

**Comment by allenhumphreys** [comment]
@kushagraThapar Our team has reached a point where we really need Cosmos to support the latest version of netty.

**Comment by kushagraThapar** [comment]
> @kushagraThapar Our team has reached a point where we really need Cosmos to support the latest version of netty.

@allenhumphreys - which version would that be ?

**Comment by allenhumphreys** [comment]
> > @kushagraThapar Our team has reached a point where we really need Cosmos to support the latest version of netty.
> 
> @allenhumphreys - which version would that be ?

Sorry for not being clear. We’d like to upgrade to spring boot 2.4, which means we need a version of Cosmos that is compatible with reactor netty v1.0.0 or higher. Netty had some minor API breakages going from the 0.9 to 1.0 version train. This GitHub issue/pr seems to be addressing support across many SDKs, I just wanted to register our desire to keep our application framework up to date.

**Comment by kushagraThapar** [comment]
> > > @kushagraThapar Our team has reached a point where we really need Cosmos to support the latest version of netty.
> > 
> > 
> > @allenhumphreys - which version would that be ?
> 
> Sorry for not being clear. We’d like to upgrade to spring boot 2.4, which means we need a version of Cosmos that is compatible with reactor netty v1.0.0 or higher. Netty had some minor API breakages going from the 0.9 to 1.0 version train. This GitHub issue/pr seems to be addressing support across many SDKs, I just wanted to register our desire to keep our application framework up to date.

Makes sense, we are trying the same, its just that we are seeing some issues with latest reactor-netty connection pool update. I am working with reactor-netty team and trying to get a repro to fix the issue from their side. Once we are all clear, we will support spring boot 2.4

**Comment by Saljack** [comment]
@kushagraThapar Can you link here these issues in reactor-netty? It can be useful when there will be an issue after upgrade. Everyone will know what to watch out for. We can track a progress.

**Comment by kushagraThapar** [comment]
> @kushagraThapar Can you link here these issues in reactor-netty? It can be useful when there will be an issue after upgrade. Everyone will know what to watch out for. We can track a progress.

I am still working on a repro, but here is the conversation I am having with reactor-netty folks - https://gitter.im/reactor/reactor-netty?at=5ffe0fbedc2e4809aa1adc8b

**Pull request closed without merging**